docs/proposals/parameterized-config-management-plugins.md

```diff
@@ -0,0 +1,502 @@
+kind: ConfigManagementPlugin
+metadata:
+  name: cmp-plugin
+spec:
+  version: v1.0
+  generate:
+    command: ["example.sh"]
+  discover:
+    fileName: "./subdir/s*.yaml"
+  # NEW KEY
+  parameters:
+    # The declarative announcement follows the parameters announcement schema. This is where a parameter description
+    # should go if it applies to all apps for this CMP.
+    announcement:
+      - name: example
+    # The (optional) generated announcement is combined with the declarative announcement (if present).
+    command: ["example-params.sh"]
+```
+
+The currently-configured parameters (if there are any) will be communicated to both `generate.command` and 
+`parameters.command` via an `ARGOCD_PARAMETERS` environment variable. The parameters will be encoded according to the 
+[parameters serialization format](#parameters-manifest--parameters-serialization-format) defined below.
+
+Passing the parameters to the `parameters.command` will allow configuration of parameter discovery. For example:
+
+```yaml
+plugin:
+  parameters:
+    main:
+      - name: ignore-helm-charts
+        value: '["chart-a", "chart-b"]'
+```
+
+#### Parameters manifest / parameters serialization format
+
+Parameters announcements should be produced by the CMP as JSON. Use JSON instead of YAML because the tooling is better
+(native JSON libraries, StackOverflow answers about jq, etc.).
+
+Parameters should be set in the manifest as a map of section names to parameter name/value pairs. YAML is used because
+it's easy to read/manipulate in an editor when modifying an Application manifest. We partition by section name so that
+the manifest, to the extent possible, is laid out similarly to the UI.
+
+```yaml
+plugin:
+  parameters:
+    main:
+      - name: values-files
+        value: '["values.yaml"]'
+    Helm Parameters:
+      - name: image
+        values: some.repo:tag
+```
+
+**Note:** I'm not sure whether CRDs allow Map<string, list> types. If not, we should consider flattening the schema to
+a list of objects, each object having a `section` field.
+
+Parameters should be communicated _to_ the CMP as JSON in the same schema as is used in the Application manifest.
+JSON might be a surprising choice considering parameters are represented in the manifest as YAML. But I think JSON makes 
+sense because 1) it's used for parameters announcements (consistency is good) and 2) JSON tooling is better.
+
+#### Parameter definition schema
+
+A parameter definition is an object with following schema:
+
+```go
+type ParameterDefinition struct {
+	// Name is the name of a parameter. (required)
+	Name string `json:"name"`
+	// Type is the type of the parameter. This determines the schema of `uiConfig` and how the UI presents the 
+	// parameter. (default is `string)
+	Type string `json:"type"`
+	// UiConfig is a stringified JSON object containing information about how the UI should present the parameter. 
+	// (default is `{}`)
+	UiConfig string `json:"uiConfig"`
```

Comment:
I realize we want to keep the `uiConfig` open ended, but shouldn't we come up with an initial set of supported interface descriptions for it?  It feels to me that it's more than just an implementation detail.

Reply:
@jopit maybe... If there are any already-existing UI features we need to support, then I'd recommend defining that part of the `uiConfig` spec now. I haven't seen any features that wouldn't already be covered by `type`, but I'll take another look.

Aside: do you think `UiConfig` should be a generic JSON object instead of a stringified JSON object? If we know we want JSON, I'm not sure there's a reason to require it be stringified.

Reply:
Just to clarify things, my suggestion was to maybe use [json.RawMessage](https://pkg.go.dev/encoding/json#RawMessage) instead of `string`.

Reply:
I don't know of a reason it needs to be stringified.

The example down below shows the `enum` type needs a `uiConfig`

`types` like the following probably don't need `uiConfig`
- boolean
- string
- password

Would we want to define types and corresponding uiConfigs for things like the following?
- something like an enum, but it lets you select more than one from a list (build-in helm values file uses this), not sure what a good name for it is
- as above, but also lets you specify a value not on the list